File: tools/analyze_data.py
import json, os, statistics
from collections import Counter, defaultdict
from datetime import datetime, date

DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'data')

files = [
    'flights.json','crew.json','rosters.json','duty_clocks.json',
    'reserve_pool.json','certifications.json','rules.json','costs.json',
    'risk_signals.json','scenarios.json','questions.json'
]

def load(fname):
    p = os.path.join(DATA_DIR, fname)
    try:
        with open(p) as fh:
            return json.load(fh)
    except Exception as e:
        print(f"ERROR loading {fname}: {e}")
        return None

flights = load('flights.json') or []
crew = load('crew.json') or []
rosters = load('rosters.json') or {"pairings": []}
duty_clocks = load('duty_clocks.json') or []
reserves = load('reserve_pool.json') or []
certs = load('certifications.json') or []
rules = load('rules.json') or {}
costs = load('costs.json') or {}
risks = load('risk_signals.json') or []
scenarios = load('scenarios.json') or []
questions = load('questions.json') or []

print('=== High-level counts ===')
print('flights:', len(flights))
print('crew:', len(crew))
print('pairings:', len(rosters.get('pairings', [])))
print('duty_clocks:', len(duty_clocks))
print('reserves:', len(reserves))
print('certifications:', len(certs))
print('scenarios:', len(scenarios))
print('questions:', len(questions))

# Crew distributions
by_base = Counter(c.get('base') for c in crew)
by_rank = Counter(c.get('rank') for c in crew)
ratings_counter = Counter()
for c in crew:
    for r in c.get('ratings', []): ratings_counter[r]+=1

print('\n=== Crew distributions ===')
print('by base:', by_base.most_common())
print('by rank:', by_rank.most_common())
print('ratings counts:', ratings_counter.most_common())

# Duty clocks stats
d7 = [c.get('duty_hours_7d',0.0) for c in duty_clocks]
f28 = [c.get('flight_hours_28d',0.0) for c in duty_clocks]
if d7:
    print('\n=== Duty/Flight hours (7d/28d) ===')
    print('duty 7d: min', min(d7), 'max', max(d7), 'mean', round(statistics.mean(d7),2))
if f28:
    print('flight 28d: min', min(f28), 'max', max(f28), 'mean', round(statistics.mean(f28),2))

# Top busiest routes
route_counts = Counter()
route_seats = Counter()
for f in flights:
    route = f['dep_station'] + '->' + f['arr_station']
    route_counts[route]+=1
    route_seats[route]+= f.get('seats',0)
print('\n=== Top routes ===')
for r,cnt in route_counts.most_common(10):
    print(r, 'flights=', cnt, 'total_seats=', route_seats[r])

# Pairing lengths and FDP
pairing_days = rosters.get('pairings', [])
pairing_sector_counts = []
pairing_duty_hours = []
for p in pairing_days:
    for d in p.get('days', []):
        num_sectors = len(d.get('flights', []))
        pairing_sector_counts.append(num_sectors)
        # compute duty hours if present
        try:
            rep = datetime.strptime(d['report_utc'], '%Y-%m-%dT%H:%M:%SZ')
            rel = datetime.strptime(d['release_utc'], '%Y-%m-%dT%H:%M:%SZ')
            dh = (rel - rep).total_seconds()/3600.0
            pairing_duty_hours.append(dh)
        except:
            pass

print('\n=== Pairing stats ===')
print('avg sectors per pairing-day:', round(statistics.mean(pairing_sector_counts),2) if pairing_sector_counts else 0)
print('avg duty hours per pairing-day:', round(statistics.mean(pairing_duty_hours),2) if pairing_duty_hours else 0)

# Certifications expiring soon relative to today
as_of = date(2026,9,5)
expiring_30 = []
expiring_90 = []
for c in certs:
    try:
        vt = datetime.fromisoformat(c['valid_to']).date()
        days = (vt - as_of).days
        if days <= 30:
            expiring_30.append((c['crew_id'], c['cert_type'], c['valid_to']))
        if days <= 90:
            expiring_90.append((c['crew_id'], c['cert_type'], c['valid_to']))
    except:
        pass
print('\n=== Certifications expiring ===')
print('<=30 days:', len(expiring_30))
print('<=90 days:', len(expiring_90))
if expiring_30:
    print('sample expiring <=30d:', expiring_30[:5])

# Reserve on-call window start times distribution
start_hours = Counter()
for r in reserves:
    try:
        s = r.get('oncall_window_utc', {}).get('start') or r.get('oncall_start') or r.get('oncall_start_time')
        if s:
            h = int(s.split(':')[0])
            start_hours[h]+=1
    except:
        pass
print('\n=== Reserve on-call window start hour distribution ===')
print(start_hours.most_common())

# Risk signals: top-risk flights or pairings
risk_by_flight = Counter()
for r in risks:
    if 'flight_id' in r:
        risk_by_flight[r['flight_id']]+= r.get('score',1)
print('\n=== Risk signals ===')
print('top risky flights sample:', risk_by_flight.most_common(10)[:10])

# Questions tiers distribution
tiers = Counter(q.get('tier') for q in questions)
print('\n=== Questions by tier ===')
print(dict(tiers))

# Engineered edge cases detection
engineered = [cid for cid in ('C-1042','C-2087','C-2210','C-3305','C-3310') if any(c.get('crew_id')==cid for c in crew)]
print('\n=== Engineered/known IDs present ===')
print(engineered)

# simple heuristic patterns
patterns = []
# 1. Many flights originate/terminate at BLR
blr_count = sum(1 for f in flights if f['dep_station']=='BLR' or f['arr_station']=='BLR')
if blr_count / max(1,len(flights)) > 0.6:
    patterns.append('BLR hub dominates schedule (>60% flights)')
# 2. A320 vs ATR proportions
ac_types = Counter(f.get('aircraft_type') for f in flights)
patterns.append('Aircraft mix: ' + ', '.join(f"{k}:{v}" for k,v in ac_types.items()))

print('\n=== Discovered patterns (heuristic) ===')
for p in patterns:
    print('-', p)

print('\nAnalysis complete.')

File: tools/analyze_data.js
const fs = require('fs');
const path = require('path');
const DATA_DIR = path.join(__dirname, '..', 'data');
const read = (f) => {
  try { return JSON.parse(fs.readFileSync(path.join(DATA_DIR, f), 'utf8')) } catch(e){ console.error('ERR', f, e.message); return null }
}
const flights = read('flights.json') || [];
const crew = read('crew.json') || [];
const rosters = read('rosters.json') || {pairings: []};
const duty_clocks = read('duty_clocks.json') || [];
const reserves = read('reserve_pool.json') || [];
const certs = read('certifications.json') || [];
const risks = read('risk_signals.json') || [];
const scenarios = read('scenarios.json') || [];
const questions = read('questions.json') || [];
const costs = read('costs.json') || {};

console.log('=== High-level counts ===');
console.log('flights:', flights.length);
console.log('crew:', crew.length);
console.log('pairings:', rosters.pairings ? rosters.pairings.length : 0);
console.log('duty_clocks:', duty_clocks.length);
console.log('reserves:', reserves.length);
console.log('certifications:', certs.length);
console.log('scenarios:', scenarios.length);
console.log('questions:', questions.length);

const Counter = () => { const map = new Map(); return {
  add(k,v=1){ map.set(k, (map.get(k)||0)+v) },
  entries(){ return Array.from(map.entries()).sort((a,b)=>b[1]-a[1]) }
}};

const by_base = Counter();
const by_rank = Counter();
const ratings = Counter();
crew.forEach(c=>{ by_base.add(c.base); by_rank.add(c.rank); (c.ratings||[]).forEach(r=>ratings.add(r)) });
console.log('\n=== Crew distributions ===')
console.log('by base:', by_base.entries().slice(0,10));
console.log('by rank:', by_rank.entries());
console.log('ratings:', ratings.entries());

const d7 = duty_clocks.map(c=>c.duty_hours_7d||0).filter(x=>typeof x==='number');
const f28 = duty_clocks.map(c=>c.flight_hours_28d||0).filter(x=>typeof x==='number');
const stats = arr=>({min: Math.min(...arr), max: Math.max(...arr), mean: arr.reduce((a,b)=>a+b,0)/arr.length});
if(d7.length) console.log('\n=== Duty/Flight hours ===\n', stats(d7));
if(f28.length) console.log(stats(f28));

const route_counts = Counter(); const route_seats = new Map();
flights.forEach(f=>{ const r = f.dep_station+'->'+f.arr_station; route_counts.add(r); route_seats.set(r, (route_seats.get(r)||0)+ (f.seats||0)) });
console.log('\n=== Top routes ==='); console.log(route_counts.entries().slice(0,10).map(([r,c])=>[r,c, route_seats.get(r)]));

// pairing stats
let pairing_sector_counts = [], pairing_duty_hours = [];
(rosters.pairings||[]).forEach(p=>{ p.days.forEach(d=>{ pairing_sector_counts.push(d.flights.length); try{ const rep = new Date(d.report_utc); const rel = new Date(d.release_utc); pairing_duty_hours.push((rel-rep)/3600000); }catch(e){} })});
if(pairing_sector_counts.length) console.log('\n=== Pairing stats ===\n avg sectors:', pairing_sector_counts.reduce((a,b)=>a+b,0)/pairing_sector_counts.length);
if(pairing_duty_hours.length) console.log('avg duty hours:', pairing_duty_hours.reduce((a,b)=>a+b,0)/pairing_duty_hours.length);

const as_of = new Date('2026-09-05');
const exp30 = certs.filter(c=>{ try{ return (new Date(c.valid_to)-as_of)/(1000*60*60*24) <=30 }catch(e){return false}});
const exp90 = certs.filter(c=>{ try{ return (new Date(c.valid_to)-as_of)/(1000*60*60*24) <=90 }catch(e){return false}});
console.log('\n=== Certs expiring ===\n<=30:', exp30.length, '\n<=90:', exp90.length, exp30.slice(0,5));

const start_hours = Counter();
reserves.forEach(r=>{ const s = (r.oncall_window_utc && r.oncall_window_utc.start) || r.oncall_start || r.oncall_start_time; if(s) start_hours.add(parseInt(s.split(':')[0],10)) });
console.log('\n=== Reserve start hours ===', start_hours.entries());

const risk_by_flight = Counter(); risks.forEach(r=>{ if(r.flight_id) risk_by_flight.add(r.flight_id, r.score||1) });
console.log('\n=== Risk signals sample ===', risk_by_flight.entries().slice(0,10));

const tiers = Counter(); questions.forEach(q=>tiers.add(q.tier)); console.log('\n=== Question tiers ===', tiers.entries());

const engineered = ['C-1042','C-2087','C-2210','C-3305','C-3310'].filter(id=> crew.some(c=>c.crew_id===id));
console.log('\n=== Engineered IDs present ===', engineered);

// heuristics
let blr_count = flights.filter(f=>f.dep_station==='BLR' || f.arr_station==='BLR').length;
if(blr_count/flights.length>0.6) console.log('\nPattern: BLR hub dominates schedule (>60% flights)');
const ac = new Map(); flights.forEach(f=> ac.set(f.aircraft_type, (ac.get(f.aircraft_type)||0)+1)); console.log('Aircraft mix:', Array.from(ac.entries()));

console.log('\nAnalysis complete.');
